From 3766e545d13b4facf448c8364518b61fe116d621 Mon Sep 17 00:00:00 2001
From: Aaradhana Sahu <quic_aarasahu@quicinc.com>
Date: Sat, 13 Apr 2024 15:19:24 +0530
Subject: [PATCH] hostapd: Fix set channel failure for multi-link AP

When we set channel for multi-link AP, it fails get EBUSY error
from driver because we fetch link_id from frequency params and
fill NL80211_ATTR_MLO_LINK_ID, but this link_id value is default
value which is 0, when we try to set channel again due to bandwidth
change.

This issue can be easily seen during color change and AWGN events.

Fix this issue by filling link_id value of frequency params from
hostapd data.

Signed-off-by: Aaradhana Sahu <quic_aarasahu@quicinc.com>
---
 src/ap/beacon.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/src/ap/beacon.c b/src/ap/beacon.c
index 70b17da..27fa1f7 100644
--- a/src/ap/beacon.c
+++ b/src/ap/beacon.c
@@ -2772,6 +2772,11 @@ static int __ieee802_11_set_beacon(struct hostapd_data *hapd)
 				    iconf->center_freq_device) == 0)
 		params.freq = &freq;
 
+#ifdef CONFIG_IEEE80211BE
+	if (params.freq && hapd->conf->mld_ap)
+		params.freq->link_id = hapd->mld_link_id;
+#endif
+
 	params.allowed_freqs = NULL;
 #ifdef CONFIG_DRIVER_NL80211_QCA
 	for (i = 0; i < hapd->iface->num_hw_features; i++) {
-- 
2.34.1

